### ISSUE TITLE
Memory Error

### ISSUE THREAD

**Issue by skipjack** [Munter/hyperlink#117]
> From our discussion on webpack/webpack.js.org#1194...

We've been seeing an error in our build process over at webpack/webpack.js.org which seems to be stemming from this package:

![image](https://cloud.githubusercontent.com/assets/8988697/25762863/70a69e40-31ae-11e7-8d11-1f4f06353bd3.png)

See this [travis build](https://travis-ci.org/webpack/webpack.js.org/builds/228929474?utm_source=github_status&utm_medium=notification) for the full log.

**Comment by skipjack** [comment]
Just quoting @papandreou from webpack/webpack.js.org for reference:

> Yeah, that's how V8 dies when the JavaScript heap exceeds the maximum allowed size (even after performing a full GC). The maximum size is specified using the `node --max-old-space-size=<numMegaBytes>` switch and defaults to 2000 (2 GB).

> That could either indicate that the combined size of the assets that hyperlink is loading has increased so it doesn't fit anymore, or that there is a bug that means it'll run off and follow external relations/redirects infinitely.

> As the first thing I would try invoking it as `node --max-old-space-size=5000 hyperlink ...` to rule out the latter.

@Munter I'm trying the `node --max-old-space...` thing now to see if that does anything. I may have to take out link checking temporarily on webpack.js.org until this is fixed (as currently, all builds fail). No rush, but please keep us posted if you make progress on this.

**Comment by skipjack** [comment]
Ah so actually running it as suggested...

``` bash
node --max-old-space-size=5000 ./node_modules/.bin/hyperlink ...
```

does fix the issue, all be it somewhat hackily, but at least this might be clue as to what the underlying issue is?

**Comment by Munter** [comment]
So the problem with large memory consumption comes from Assetgraph keeping every visited asset in memory for the duration of the run. If the graph is huge, contains large binaries like images or videos, the memory consumption will grow very quickly.

Possible way to mitigate this:
- Avoid populating asset types that are known to be large and not have outgoing relations: Assets that inherit from `Image`, `Flash`, `Asset` (non-modeled asset type, no outgoing relations)
- Reduce memory footprint of the graph by clearing out `.text`, `.rawSrc`, `.parseTree` when it is known that an asset has been fully processed and its `.outgoingRelations` have been queued in the `populate`-transform.

Both approaches require a bit of work.

Avoiding population of large binary asset is probably the biggest bang for the buck, and the easiest to implement. We'd have to fine tune the population query a bit and then round up those internal unpopulated assets and do a manual check on them without actually loading them. For http assets this part is already handled, since this is the approach I took with checking external outgoing links. For runs on local disk there would have to be some new code path that does an `fs.stat` or similar to replace the http calls.

Cleaning out the sources of the already populated graph along the way is slightly more problematic. There are no events or hooks that describe that population of an assets outgoing relation has been queued. Further more there is no built-in method to unload an asset without also killing relations and the asset itself, so there would have to be a manual deletion of internal asset data. We could either add some extra events in Assetgraph core to support this use case or we'd have to basically re-implement the `populate` transform for the special case.

@papandrou do you agree with the above assessment? Is this use case something that should be possible within the scope of Assetgraph core (the events at least)?

**Comment by Munter** [comment]
There might be further memory and performance wins by starting external asset checks out with an `HTTP HEAD` instead of `HTTP GET`, then fall back to `HTTP GET` if the returned status indicates non-support of `HEAD`. The bandwidth use could be reduced considerably. Or we could do an `HTTP GET` with a stream and just hang up as soon as the relevant data is available

**Comment by Munter** [comment]
A memory reduction has been put in place, which reduces the footprint quite a bit. Tested to work on the latest wbepack.js.org.

Released in 2.8.0
